docs/proposals/client-package-structure.md

```diff
@@ -0,0 +1,153 @@
+subsequent layers.
+
+Tentative name: "Transport".
+
+Action items:
+* Split out of current client package into a new package. (@krousey)
+
+### RESTClient/request.go
+
+RESTClient consumes a Transport, a group, and a version, and produces something
+that implements the interface currently in request.go. That is, with a
+RESTClient, you can write chains of calls like: c.Get().Path(p).Param("name",
+"value").Do()
+
+The group and version may be empty when constructing a RESTClient. This is valid
+for executing discovery commands. The group and version may be overridable with
+a function call.
+
+RESTClients may have semantic information built in. For example, if the
+query parameter changes from "labels" to "label-selector", when you call the
+RESTClient construction logic, it should give you a RESTClient which uses the
+appropriate query parameter when you call the "LabelSelector()" method.
+
+3rd parties such as open shift may provide their own implementation, but it
+should conform to the same interface as the standard RESTClient.
+
+Action items:
+* Split out of the current client package. (@krousey)
+* Convert to an interface (currently, it's a struct).
+
+### Mux layer
+
+The client muxer layer has a map of group/version to cached RESTClient, and
+knows how to construct a new RESTClient in case of a cache miss.
+
+Tentative name: MetaClient
+
+Action items:
+* Move client cache out of kubectl libraries into a more general home.
+
+### High-level: Individual typed
+
+Our current high-level client allows you to write things like
+`c.Pods("namespace").Create(p)`; we will insert a level for the group.
+
+That is, the system will be:
+`client.GroupName().NamespaceSpecifier().RESTAction()`
+
+The types returned/consumed by such functions will be e.g. api/v1, NOT the
+current version inspecific types.
+
+We will provide this structure for each version of each group. It is infeasible
+to do this manually, so we will generate this. The generator will accept both
+swagger and the ordinary go types.
+
+Typed clients will be constructabale given a MetaClient.
+
+Action items:
+* Move current typed clients into new directory structure (described below)
+* Finish client generation logic. (@lavalamp)
+
+#### High-level, typed: Discovery
+
+A `DiscoveryClient` is necessary to discover the api groups, versions, and
+resources a server supports. It's constructable given a RESTClient. It is
+consumed by both the MetaClient and users who want to iterate over groups,
+versions, or resources. (Example: namespace controller.)
+
+Action items:
+* Refactor existing functions to present a sane interface, as close to that
+  offered by the other typed clients as possible. (@caeserxuchao)
+
+### High-level: Client Sets
+
+Because there will be multiple groups with multiple versions, we will provide an
+aggregation layer that combines multiple typed clients in a single object. We do
+this to constrain the testing matrix.
+
```

Comment:
How does the constrain the testing matrix?  What code needs this abstraction?  Is it non-Kube golang client consumers?  

If so, I assume it would be in its own package?